src/main/scala/io/iohk/ethereum/network/PeerManagerActor.scala

```diff
@@ -134,9 +143,15 @@ class PeerManagerActor(
   }
 
   private def maybeConnectToDiscoveredNodes(connectedPeers: ConnectedPeers, nodes: Set[Node]): Unit = {
-    val nodesToConnect = nodes
+    val discoveredNodes = nodes
       .filter(connectedPeers.canConnectTo)
-      .take(connectedPeers.outgoingConnectionDemand)
+
+    val nodesToConnect = discoveredNodes
+      .filterNot(n => triedNodes.contains(n.id)) match {
+      case seq if seq.size >= connectedPeers.outgoingConnectionDemand =>
+        seq.take(connectedPeers.outgoingConnectionDemand)
+      case _ => discoveredNodes.take(connectedPeers.outgoingConnectionDemand)
```

Comment:
I am no sure if this will be a problem, but this means that if network size is less than `triedNodes.maxSize`, then after first traversal we are back to our old way of checking/establishing connections, and if network size is much larger then we may still not try all nodes as they will be constantly changing in `triedNodes` cache and in `blacklist`. So in summary, I am not sure how much this will help, and the best way to find out is run mantis several times on etc mainnet and mordor to check it out.

Reply:
the logic behind that is
- if the network size is small - then it's better to retry previously failed nodes than not do anything at all if we don't have any new nodes to try
- if the network size is large - we should try only new nodes until the moment the cache is full and we start replacing old values with new ones. 

Do you think it makes sense to review the max size of the cache?

Reply:
I agree with this logic, and probably it is enough for small network. 

But in case of large network, we may still not try all peer provided by discovery. Lets say the network have 8k peers (which after some time is not impossible) and current max cache size is 4k. Then after cache gets full, our situation is basically the same as currently i.e we will have 4k peers to try and will have no way to be certain we tried them all.

as for reviewing cache size it may help, but i am not sure there is way to derive optimal value as network size can size dynamically.

In general I like this apprach becouse it solves problem in most of cases in simple way, but the question is this enough are what are benefits in comparission to baseline.